### ISSUE TITLE
Does an empty struct have the same layout as the empty tuple?

### ISSUE THREAD

**Issue by joshlf** [rust-lang/unsafe-code-guidelines#315]
If I'm consuming the following API:

```rust
pub struct Foo;
```

...am I guaranteed that `Foo` has the same layout as `()`? Or do I need an explicit `repr` attribute?

**Comment by thomcc** [comment]
IIUC it's not guaranteed, but it's hard to imagine it not being the case. I remember asking about this before (admittedly, my case was `#[repr(align(n))] struct Foo;`, which is slightly different), and being told I'd need to write an RFC to guarantee it, but that it should be uncontroversial (unfortunately, I have too many unfinished RFCs in my RFC repo already to want to finish this).

Just to head it off, since you implied it via a question: adding a `#[repr(C)]` seems like a bad idea, given that `struct Foo {}` in C has a size of 1. That said, currently, I believe it has a size of zero in Rust, which is arguably a bug.

To answer your second question: unless it's `#[non_exhaustive]`, I don't believe there's any API-compatible change they can make that will affect the layout. (I guess in theory they could add an explicit repr, which if the "arguable bug" I mentioned in the previous paragraph is ever fixed, would impact the layout.)

**Comment by Ixrec** [comment]
I'm fairly sure it's not a bug that empty types are zero-sized. IIUC it's a significant difference from C and C++, but Rust has pretty thorough support for ZSTs (zero-sized types), and that's an important part of making some types "zero-overhead" in practice. Even in C++, letting things be zero-sized is often considered a desirable but illegal optimization, [and as of C++20 you can sometimes opt-in to it](https://en.cppreference.com/w/cpp/language/attributes/no_unique_address).

I agree that `#[repr(C)]` is not the solution here. 

```rust
#[repr(transparent)]
struct Foo(());
```
^would be my first guess for what to use today, assuming "same layout as `()`" is literally what you need.

Either way, I don't think most people would expect type layout to be part of a crate's public API by default. If you're trying to say "I promise you can transmute my crate's type to `()` and I won't break that without a major version bump", rather than try to promise this directly I'd add a public API function to do that and let the actual transmute live in your library code.

(although I'm guessing `()` was just an example)

**Comment by thomcc** [comment]
> I'm fairly sure it's not a bug that empty types are zero-sized. IIUC it's a significant difference from C and C++, but Rust has pretty thorough support for ZSTs (zero-sized types), and that's an important part of making some types "zero-overhead" in practice.

I meant that it's arguably a bug that  `#[repr(C)] struct Foo {}` doesn't match C's `struct Foo {};`'s repr.

**Comment by bobrippling** [comment]
> given that struct Foo {} in C has a size of 1

> I guess it might be implementation defined to have an empty struct in C though

Yeah, I expect the reason this isn't supported currently is because it's different in C and C++:
- An empty struct/class in C++ has size 1 (but might be combined with empty base classes)
- An empty struct in C is [not allowed](https://cigix.me/c17#6.2.5.p20) (_"nonempty set of member objects"_), but supported as an extension in GNU C, where it has size 0
